File: src/services/TwStockService.js
const axios = require('axios');
const logger = require('../utils/logger');
const config = require('../config/config');
class TwStockService {
    constructor() {
        this.quoteCache = {}; // 建立台股專屬快取
        setInterval(() => {
            const now = Date.now();
            for (const sym in this.quoteCache) {
                if (now - this.quoteCache[sym].timestamp > config.CACHE_MAX_AGE) {
                    delete this.quoteCache[sym];
                }
            }
        }, config.CACHE_CLEAN_INTERVAL);
    }
    getHeaders() {
        const agents = [
            'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36',
            'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/119.0.0.0 Safari/537.36',
            'Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:109.0) Gecko/20100101 Firefox/115.0'
        ];
        return {
            'User-Agent': agents[Math.floor(Math.random() * agents.length)],
            'Accept-Language': 'zh-TW,zh;q=0.9,en-US;q=0.8,en;q=0.7'
        };
    }

    async searchTwse(query) {
        try {
            const cleanQuery = query.replace(/\.TW[O]?$/i, '');
            const url = `https://mis.twse.com.tw/stock/api/getStockNames.jsp?n=${encodeURIComponent(cleanQuery)}&lang=zh_tw`;
            const { data } = await axios.get(url, { headers: this.getHeaders() });

            if (data && data.datas && Array.isArray(data.datas)) {
                return {
                    quotes: data.datas.map(item => {
                        const suffix = item.t === 'otc' ? '.TWO' : '.TW';
                        return {
                            symbol: `${item.c}${suffix}`,
                            shortname: item.n,
                            longname: item.n,
                            exchange: item.t === 'otc' ? 'TWO' : 'TW',
                            isYahooFinance: true
                        };
                    })
                };
            }
        } catch (e) { logger.warn(`[TwStock] TWSE 搜尋失敗: ${e.message}`); }
        return null;
    }

    async fetchTwseMonthChart(symbol) {
        try {
            const isOTC = symbol.includes('.TWO');
            const stockNo = symbol.split('.')[0];
            const date = new Date();
            const yyyy = date.getFullYear();
            const mm = String(date.getMonth() + 1).padStart(2, '0');
            const rocYear = yyyy - 1911;

            let url = isOTC
                ? `https://www.tpex.org.tw/web/stock/aftertrading/daily_trading_info/st43_result.php?l=zh-tw&d=${rocYear}/${mm}&stk_no=${stockNo}`
                : `https://www.twse.com.tw/exchangeReport/STOCK_DAY?response=json&date=${yyyy}${mm}01&stockNo=${stockNo}`;

            logger.info(`[TwStock] 抓取台股官方本月 K 線: ${stockNo}`);
            const { data } = await axios.get(url, { headers: this.getHeaders(), timeout: 10000 });

            let rawData = [];
            if (isOTC && data.aaData) rawData = data.aaData;
            else if (!isOTC && data.data) rawData = data.data;

            if (!rawData || rawData.length === 0) throw new Error('本月尚無交易數據或代碼錯誤');

            const formatted = rawData.map(row => {
                const parts = row[0].split('/');
                const gregorianYear = parseInt(parts[0]) + 1911;
                return {
                    time: `${gregorianYear}-${parts[1]}-${parts[2]}`,
                    open: parseFloat(row[3].replace(/,/g, '')),
                    high: parseFloat(row[4].replace(/,/g, '')),
                    low: parseFloat(row[5].replace(/,/g, '')),
                    close: parseFloat(row[6].replace(/,/g, ''))
                };
            }).filter(item => !isNaN(item.open) && !isNaN(item.close));

            return formatted;
        } catch (e) {
            logger.error(`[TwStock] 官方 K 線抓取失敗 (${symbol}): ${e.message}`);
            throw new Error('無法取得證交所資料，請稍後再試');
        }
    }

    async fetchFugleChart(symbol, apiKey) {
        try {
            const stockNo = symbol.split('.')[0];
            const start = new Date();
            start.setMonth(start.getMonth() - 6);
            const fromDate = start.toISOString().split('T')[0];
            const toDate = new Date().toISOString().split('T')[0];

            logger.info(`[Fugle] 抓取富果近半年 K 線: ${stockNo}`);
            const url = `https://api.fugle.tw/marketdata/v1.0/stock/historical/candles/1d?symbol=${stockNo}&from=${fromDate}&to=${toDate}`;
            const { data } = await axios.get(url, { headers: { 'X-API-KEY': apiKey }, timeout: 10000 });

            if (!data.data || data.data.length === 0) throw new Error('查無富果 K 線數據');

            const formatted = data.data.map(item => ({
                time: item.date,
                open: item.open,
                high: item.high,
                low: item.low,
                close: item.close
            }));

            formatted.sort((a, b) => new Date(a.time) - new Date(b.time));
            return formatted;
        } catch (e) {
            logger.error(`[Fugle] API 抓取失敗 (${symbol}): ${e.response?.data?.message || e.message}`);
            if (e.response?.status === 401) throw new Error('富果 API Key 錯誤或無效');
            throw new Error('無法取得富果資料，請確認 API 狀態');
        }
    }
    // 🚀 [NEW] 富果盤中即時報價引擎
    async fetchFugleQuotes(symbols, apiKey) {
        if (!symbols || symbols.length === 0 || !apiKey) return {};
        const results = {};
        const now = Date.now();
        const symbolsToFetch = [];

        const twDate = new Date(new Date().toLocaleString('en-US', { timeZone: 'Asia/Taipei' }));
        const day = twDate.getDay();
        const h = twDate.getHours();
        const m = twDate.getMinutes();
        const isTradingHours = (day >= 1 && day <= 5) && ((h === 9 && m >= 0) || (h >= 10 && h <= 12) || (h === 13 && m <= 30));

        // 如果是盤中，快取 15 秒 (省額度但保持流暢)
        // 如果已收盤或假日，快取 1 小時 (3600000毫秒，極致省額度！)
        const cacheTTL = isTradingHours ? config.CACHE_TRADING_OPEN : config.CACHE_TRADING_CLOSED;

        symbols.forEach(sym => {
            const cached = this.quoteCache[sym];
            // 使用智慧計算出來的 TTL 來防護
            if (cached && (now - cached.timestamp < cacheTTL)) {
                results[sym] = cached.data;
            } else {
                symbolsToFetch.push(sym);
            }
        });

        if (symbolsToFetch.length > 0) {
            // 富果 API 一次只能查一檔，所以我們用 Promise.all 並行抓取
            await Promise.allSettled(symbolsToFetch.map(async (sym) => {
                try {
                    const stockNo = sym.split('.')[0];
                    const url = `https://api.fugle.tw/marketdata/v1.0/stock/intraday/quote/${stockNo}`;
                    const { data } = await axios.get(url, { headers: { 'X-API-KEY': apiKey } });

                    // [修正] 富果 v1.0 回傳的 JSON 直接就在 data 裡面，不用往深層剝
                    if (data && data.symbol) {
                        const q = data;

                        // 將富果的欄位精準對應到我們的系統格式
                        const quoteData = {
                            price: q.lastPrice || q.closePrice || q.previousClose || 0,
                            prevClose: q.previousClose || 0,
                            high: q.highPrice || 0,
                            low: q.lowPrice || 0,
                            currency: 'TWD',
                            marketStatus: q.isClose ? 'CLOSED' : 'OPEN'
                        };
                        results[sym] = quoteData;
                        this.quoteCache[sym] = { data: quoteData, timestamp: now };
                    }
                } catch (e) {
                    logger.error(`[Fugle] 報價抓取失敗 (${sym}): ${e.message}`);
                }
            }));
        }
        return results;
    }
}

module.exports = TwStockService;

File: src/main/java/com/example/golden_retriever_java/service/TwStockService.java
package com.example.golden_retriever_java.service;

import com.example.golden_retriever_java.dto.ChartDataDto;
import com.example.golden_retriever_java.dto.StockQuoteDto;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.http.HttpHeaders;
import org.springframework.http.ResponseEntity;
import org.springframework.scheduling.annotation.Scheduled;
import org.springframework.stereotype.Service;
import org.springframework.web.reactive.function.client.WebClient;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.time.format.DateTimeFormatter;
import java.util.*;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.atomic.AtomicInteger;

@Service
public class TwStockService {
    private static final Logger log = LoggerFactory.getLogger(TwStockService.class);
    private final WebClient webClient;
    private final ObjectMapper objectMapper = new ObjectMapper();

    // ==========================================
    // 📊 API 額度監控
    // ==========================================
    private final AtomicInteger remainingQuota = new AtomicInteger(-1);

    public int getRemainingQuota() {
        return remainingQuota.get();
    }

    private void updateQuota(HttpHeaders headers) {
        if (headers == null) return;
        String remaining = headers.getFirst("X-RateLimit-Remaining");
        if (remaining != null) {
            try {
                int val = Integer.parseInt(remaining);
                remainingQuota.set(val);
                log.info("[Fugle Quota] 剩餘額度: {}", val);
            } catch (Exception ignored) {}
        }
    }

    // ==========================================
    // 🚀 圖表快取系統
    // ==========================================
    private static class CacheEntry<T> {
        final T data;
        final long timestamp;
        CacheEntry(T data) {
            this.data = data;
            this.timestamp = System.currentTimeMillis();
        }
    }

    private final Map<String, CacheEntry<List<ChartDataDto>>> chartCache = new ConcurrentHashMap<>();
    private static final long CACHE_DURATION = 3600000; // 快取 1 小時

    public TwStockService(WebClient webClient) {
        this.webClient = webClient;
    }

    private String resolveFugleKey(String rawKey) {
        if (rawKey == null) return "";
        return rawKey.trim();
    }

    public Map<String, StockQuoteDto> fetchFugleQuotes(List<String> symbols, String fugleKey) {
        Map<String, StockQuoteDto> results = new HashMap<>();
        String effectiveKey = resolveFugleKey(fugleKey);
        if (effectiveKey.isEmpty()) return results;

        for (String symbol : symbols) {
            try {
                String rawSymbol = symbol.split("\\.")[0];
                String url = "https://api.fugle.tw/marketdata/v1.0/stock/intraday/quote/" + rawSymbol;

                ResponseEntity<JsonNode> response = webClient.get()
                        .uri(url)
                        .header("X-API-KEY", effectiveKey)
                        .header("User-Agent", "Mozilla/5.0")
                        .retrieve()
                        .toEntity(JsonNode.class)
                        .block();

                if (response != null) {
                    updateQuota(response.getHeaders());
                    JsonNode node = response.getBody();
                    if (node != null) {
                        BigDecimal lastPrice = new BigDecimal(node.path("lastPrice").asText("0"));
                        if (lastPrice.compareTo(BigDecimal.ZERO) == 0) lastPrice = new BigDecimal(node.path("closePrice").asText("0"));
                        if (lastPrice.compareTo(BigDecimal.ZERO) == 0) lastPrice = new BigDecimal(node.path("previousClose").asText("0"));

                        StockQuoteDto dto = StockQuoteDto.builder()
                                .price(lastPrice)
                                .prevClose(new BigDecimal(node.path("previousClose").asText("0")))
                                .high(new BigDecimal(node.path("highPrice").asText("0")))
                                .low(new BigDecimal(node.path("lowPrice").asText("0")))
                                .volume(node.path("tradingVolume").asLong(0))
                                .currency("TWD")
                                .marketStatus(node.path("isClose").asBoolean(false) ? "CLOSED" : "OPEN")
                                .build();

                        // 台股額外獲取 52 週數據 (透過歷史 K 線計算)
                        List<ChartDataDto> yearData = fetchFugleChart(symbol, effectiveKey, "1y");
                        if (yearData != null && !yearData.isEmpty()) {
                            BigDecimal yHigh = yearData.stream().map(ChartDataDto::getHigh).max(BigDecimal::compareTo).orElse(lastPrice);
                            BigDecimal yLow = yearData.stream().map(ChartDataDto::getLow).min(BigDecimal::compareTo).orElse(lastPrice);
                            dto.setFiftyTwoWeekHigh(yHigh);
                            dto.setFiftyTwoWeekLow(yLow);
                            
                            long avgVol = (long) yearData.stream().mapToLong(d -> 0L).average().orElse(0); // 這裡簡化，原 DTO 沒存每日量
                            dto.setAvgVolume(avgVol);
                        }

                        results.put(symbol, dto);
                    }
                }
            } catch (Exception e) {
                log.error("[Fugle Quotes] {} 抓取失敗 | 原因: {}", symbol, e.getMessage());
            }
        }
        return results;
    }

    public List<ChartDataDto> fetchFugleChart(String symbol, String fugleKey, String period) {
        String cacheKey = "FUGLE_" + symbol + "_" + period;
        CacheEntry<List<ChartDataDto>> cached = chartCache.get(cacheKey);
        if (cached != null && (System.currentTimeMillis() - cached.timestamp < CACHE_DURATION)) {
            return cached.data;
        }

        List<ChartDataDto> chartData = new ArrayList<>();
        String effectiveKey = resolveFugleKey(fugleKey);
        if (effectiveKey.isEmpty()) return chartData;

        try {
            String rawSymbol = symbol.split("\\.")[0];
            String fromDate;
            if ("1wk".equals(period) || "1w".equals(period)) {
                fromDate = LocalDate.now().minusYears(2).format(DateTimeFormatter.ofPattern("yyyy-MM-dd"));
            } else if ("1mo".equals(period) || "1m".equals(period)) {
                fromDate = LocalDate.now().minusYears(5).format(DateTimeFormatter.ofPattern("yyyy-MM-dd"));
            } else {
                fromDate = LocalDate.now().minusMonths(6).format(DateTimeFormatter.ofPattern("yyyy-MM-dd"));
            }
            
            String toDate = LocalDate.now().format(DateTimeFormatter.ofPattern("yyyy-MM-dd"));
            String url = "https://api.fugle.tw/marketdata/v1.0/stock/historical/candles/1d?symbol=" + rawSymbol + "&from=" + fromDate + "&to=" + toDate;

            ResponseEntity<String> response = webClient.get()
                    .uri(url)
                    .header("X-API-KEY", effectiveKey)
                    .header("User-Agent", "Mozilla/5.0")
                    .retrieve()
                    .toEntity(String.class).block();

            if (response != null) {
                updateQuota(response.getHeaders());
                JsonNode root = objectMapper.readTree(response.getBody());
                JsonNode candles = root.path("data");
                
                List<ChartDataDto> dailyData = new ArrayList<>();
                if (candles.isArray()) {
                    for (JsonNode c : candles) {
                        dailyData.add(new ChartDataDto(
                                c.path("date").asText(),
                                new BigDecimal(c.path("open").asText()),
                                new BigDecimal(c.path("high").asText()),
                                new BigDecimal(c.path("low").asText()),
                                new BigDecimal(c.path("close").asText())));
                    }
                }
                dailyData.sort(Comparator.comparing(ChartDataDto::getTime));

                if ("1w".equals(period) || "1wk".equals(period)) {
                    chartData = aggregateByPeriod(dailyData, "WEEK");
                } else if ("1m".equals(period) || "1mo".equals(period)) {
                    chartData = aggregateByPeriod(dailyData, "MONTH");
                } else {
                    chartData = dailyData;
                }
                
                if (!chartData.isEmpty()) {
                    chartCache.put(cacheKey, new CacheEntry<>(chartData));
                }
            }
        } catch (Exception e) {
            log.error("[Fugle Chart] 圖表抓取失敗: {}", e.getMessage());
        }
        return chartData;
    }

    private List<ChartDataDto> aggregateByPeriod(List<ChartDataDto> dailyData, String type) {
        if (dailyData == null || dailyData.isEmpty()) return new ArrayList<>();
        List<ChartDataDto> aggregated = new ArrayList<>();
        ChartDataDto currentBar = null;
        String currentKey = "";
        DateTimeFormatter dayFormatter = DateTimeFormatter.ofPattern("yyyy-MM-dd");

        for (ChartDataDto day : dailyData) {
            LocalDate date = LocalDate.parse(day.getTime(), dayFormatter);
            String key = "WEEK".equals(type) ? date.with(java.time.DayOfWeek.MONDAY).toString() : date.withDayOfMonth(1).toString();

            if (!key.equals(currentKey)) {
                if (currentBar != null) aggregated.add(currentBar);
                currentBar = new ChartDataDto(key, day.getOpen(), day.getHigh(), day.getLow(), day.getClose());
                currentKey = key;
            } else {
                if (day.getHigh().compareTo(currentBar.getHigh()) > 0) currentBar.setHigh(day.getHigh());
                if (day.getLow().compareTo(currentBar.getLow()) < 0) currentBar.setLow(day.getLow());
                currentBar.setClose(day.getClose());
            }
        }
        if (currentBar != null) aggregated.add(currentBar);
        return aggregated;
    }

    public List<ChartDataDto> fetchTwseMonthChart(String symbol) {
        String cacheKey = "TWSE_" + symbol;
        CacheEntry<List<ChartDataDto>> cached = chartCache.get(cacheKey);
        if (cached != null && (System.currentTimeMillis() - cached.timestamp < CACHE_DURATION)) {
            return cached.data;
        }

        List<ChartDataDto> chartData = new ArrayList<>();
        try {
            String rawSymbol = symbol.split("\\.")[0];
            String dateStr = LocalDate.now().format(DateTimeFormatter.ofPattern("yyyyMMdd"));
            String url = "https://www.twse.com.tw/exchangeReport/STOCK_DAY?response=json&date=" + dateStr + "&stockNo=" + rawSymbol;
            
            JsonNode root = webClient.get()
                    .uri(url)
                    .header("User-Agent", "Mozilla/5.0")
                    .retrieve()
                    .bodyToMono(JsonNode.class)
                    .timeout(java.time.Duration.ofSeconds(5))
                    .block();
            
            if (root != null) {
                JsonNode dataNode = root.path("data");
                if (dataNode.isArray()) {
                    for (JsonNode day : dataNode) {
                        String rocDate = day.get(0).asText();
                        String[] parts = rocDate.split("/");
                        int year = Integer.parseInt(parts[0]) + 1911;
                        chartData.add(new ChartDataDto(year + "-" + parts[1] + "-" + parts[2],
                                new BigDecimal(day.get(3).asText().replace(",", "")),
                                new BigDecimal(day.get(4).asText().replace(",", "")),
                                new BigDecimal(day.get(5).asText().replace(",", "")),
                                new BigDecimal(day.get(6).asText().replace(",", ""))));
                    }
                }
            }
            if (!chartData.isEmpty()) chartCache.put(cacheKey, new CacheEntry<>(chartData));
        } catch (Exception e) { 
            log.error("[TWSE] 抓取失敗: {}", e.getMessage()); 
        }
        return chartData;
    }

    @Scheduled(fixedRate = 1800000)
    public void cleanCache() {
        long now = System.currentTimeMillis();
        chartCache.entrySet().removeIf(entry -> now - entry.getValue().timestamp > CACHE_DURATION);
    }

    private final Map<String, CacheEntry<JsonNode>> searchCache = new ConcurrentHashMap<>();
    private static final long SEARCH_CACHE_DURATION = 600000;

    public JsonNode searchTwse(String query) {
        if (query == null || query.trim().isEmpty()) return objectMapper.createObjectNode();
        String q = query.trim();
        CacheEntry<JsonNode> cached = searchCache.get(q);
        if (cached != null && (System.currentTimeMillis() - cached.timestamp < SEARCH_CACHE_DURATION)) return cached.data;

        try {
            String cleanQuery = q.replaceAll("(?i)\\.TW[O]?$", "");
            String encodedQuery = java.net.URLEncoder.encode(cleanQuery, java.nio.charset.StandardCharsets.UTF_8.toString());
            String url = "https://mis.twse.com.tw/stock/api/getStockNames.jsp?n=" + encodedQuery + "&lang=zh_tw";
            
            String res = webClient.get()
                    .uri(url)
                    .header("User-Agent", "Mozilla/5.0")
                    .retrieve()
                    .bodyToMono(String.class)
                    .timeout(java.time.Duration.ofMillis(1500))
                    .block();

            if (res != null && !res.isEmpty() && res.trim().startsWith("{")) {
                JsonNode root = objectMapper.readTree(res);
                if (root.has("datas") && root.path("datas").isArray()) {
                    var quotes = objectMapper.createArrayNode();
                    for (JsonNode item : root.path("datas")) {
                        var node = objectMapper.createObjectNode();
                        String suffix = "otc".equals(item.path("t").asText()) ? ".TWO" : ".TW";
                        node.put("symbol", item.path("c").asText() + suffix);
                        node.put("shortname", item.path("n").asText());
                        node.put("longname", item.path("n").asText());
                        node.put("exchange", suffix.substring(1));
                        quotes.add(node);
                    }
                    var result = objectMapper.createObjectNode();
                    result.set("quotes", quotes);
                    searchCache.put(q, new CacheEntry<>(result));
                    return result;
                }
            }
        } catch (Exception ignored) {}
        return objectMapper.createObjectNode();
    }

    @Scheduled(fixedRate = 3600000)
    public void cleanSearchCache() {
        long now = System.currentTimeMillis();
        searchCache.entrySet().removeIf(entry -> now - entry.getValue().timestamp > SEARCH_CACHE_DURATION);
    }

    /**
     * 抓取盤中內外盤力道
     */
    public Map<String, Object> fetchIntradayStrength(String symbol, String fugleKey) {
        Map<String, Object> result = new HashMap<>();
        String effectiveKey = resolveFugleKey(fugleKey);
        if (effectiveKey.isEmpty()) return result;

        try {
            String rawSymbol = symbol.split("\\.")[0];
            String url = "https://api.fugle.tw/marketdata/v1.0/stock/intraday/volumes/" + rawSymbol;

            ResponseEntity<JsonNode> response = webClient.get()
                    .uri(url)
                    .header("X-API-KEY", effectiveKey)
                    .header("User-Agent", "Mozilla/5.0")
                    .retrieve()
                    .toEntity(JsonNode.class)
                    .block();

            if (response != null && response.getBody() != null) {
                updateQuota(response.getHeaders());
                JsonNode volumes = response.getBody().path("volumes");
                long buyVol = 0;
                long sellVol = 0;

                if (volumes.isArray()) {
                    BigDecimal maxVol = BigDecimal.ZERO;
                    BigDecimal densePrice = BigDecimal.ZERO;

                    for (JsonNode v : volumes) {
                        long b = v.path("buyVolume").asLong(0);
                        long s = v.path("sellVolume").asLong(0);
                        buyVol += b;
                        sellVol += s;

                        BigDecimal totalVol = new BigDecimal(b + s);
                        if (totalVol.compareTo(maxVol) > 0) {
                            maxVol = totalVol;
                            densePrice = new BigDecimal(v.path("price").asText("0"));
                        }
                    }
                    result.put("dense_price", densePrice);
                    result.put("dense_vol_shares", maxVol);
                }

                long total = buyVol + sellVol;
                double buyRatio = total > 0 ? (double) buyVol / total * 100 : 50.0;
                
                result.put("buy_ratio", Math.round(buyRatio * 10.0) / 10.0);
                result.put("buy_vol", buyVol);
                result.put("sell_vol", sellVol);
                result.put("strength_msg", buyRatio > 55 ? "多頭主導" : (buyRatio < 45 ? "空頭主導" : "多空拉鋸"));
            }
        } catch (Exception e) {
            log.error("[Fugle Strength] {} 獲取失敗: {}", symbol, e.getMessage());
        }
        return result;
    }
}